### PR TITLE
Llama3 Context Parallel Fixes and Performance edits

### PR DESCRIPTION
**Pull request by pstjohn**
A collection of small performance improvements and bugfixes for llama3 CP training

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Per-sequence padding control in data processing pipelines
  * Asynchronous batch prefetching for improved training performance
  * NVIDIA Nsight Systems profiling integration for performance analysis

* **Documentation**
  * Added performance profiling guide with Nsight Systems configuration and best practices

* **Refactor**
  * Updated profiler configuration from schedule-based to explicit step-range format
  * Updated Docker base image for improved compatibility

### CODE REVIEW

**Comment by pstjohn** [comment]
@coderabbitai resolve

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 1]
```diff
@@ -234,20 +236,50 @@ class TokenPackingDataset(torch.utils.data.IterableDataset):
...
+    the remaining batch capacity.
+    """
+
+    def __post_init__(self):
+        """Validate padding configuration."""
+        if (
+            self.pad_sequences_to_be_divisible_by is not None
+            and self.max_tokens_per_batch % self.pad_sequences_to_be_divisible_by != 0
+        ):
+            logger.warning(
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:254]: so it just fills it as much as possible but maintains the divisibility factor?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 2]
```diff
@@ -234,20 +236,50 @@ class TokenPackingDataset(torch.utils.data.IterableDataset):
...
+                "Batches may not fill to exactly max_tokens_per_batch when split_samples=True.",
+                self.max_tokens_per_batch,
+                self.pad_sequences_to_be_divisible_by,
+            )
+
+    def _padded_len(self, length: int) -> int:
+        """Return the padded length of a sequence, rounding up to the nearest multiple of pad_sequences_to_be_divisible_by."""
+        if self.pad_sequences_to_be_divisible_by is None:
+            return length
+        return -(-length // self.pad_sequences_to_be_divisible_by) * self.pad_sequences_to_be_divisible_by
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:265]: whats with the `-(-X)` ?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 3]
```diff
@@ -261,15 +293,19 @@ def __iter__(self):
...
-                    tokens_in_batch = current_length - len(sample["input_ids"])
-                    # Calculate how many tokens we can fit from this sample
+                    # Calculate how many padded tokens are already in the batch
+                    tokens_in_batch = current_length - self._padded_len(len(sample["input_ids"]))
+                    # Calculate how many tokens we can fit from this sample, ensuring the
+                    # padded length doesn't exceed the remaining capacity.
                     tokens_available = self.max_tokens_per_batch - tokens_in_batch
+                    if self.pad_sequences_to_be_divisible_by is not None:
+                        d = self.pad_sequences_to_be_divisible_by
+                        tokens_available = (tokens_available // d) * d
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:303]: nit: why do we need `d` here? Can't we just use `self.pad_seq...` or is it ugly because its gonna be too long a line?

File: bionemo-recipes/models/esm2/src/esm/collator.py [hunk 4]
```diff
@@ -852,6 +932,7 @@ class BatchType(TypedDict):
     pad_between_seqs: bool
 
 
+@nvtx.annotate("collator._scatter_batch_to_cp_tp_ranks", color="green")
```
Comment (jomitchellnv) [inline bionemo-recipes/models/esm2/src/esm/collator.py:937]: The decorator at runtime does nothing if we aren't doing profiling right?

Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py:937]: yeah exactly, its just there if we're running nsys

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 3 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py:303]: probably, it's a super long varname

File: bionemo-recipes/models/esm2/src/esm/collator.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/models/esm2/src/esm/collator.py:265]: right, probably not the cleanest implementation. but `//` is a integer floor devision, `4//3=1`. So `-(-4//3)` is essentially a ceiling division, rounding up to 2
